### DISCUSSION TITLE
Directus and Firebase Auth ??

### DISCUSSION THREAD

**Discussion by dukesx** [directus/directus#4635] [Archive]
I am looking forward to developing a Blog/CMS along with a mobile app made in flutter. I am very excited about Directus, its a favorite and always the go-to choice but this time, since OAuth is my priority and handling Directus in flutter is a cumbersome work, i switched to Firebase Auth for easier integration. 

Now firebase does supply a token after authentication but my question is, will that token work with Directus ? I need to see what does Directus expect in the JWT temporary token so that i can make GraphQL calls using Bearer Token.

**Comment by rijkvanzanten** [comment]
I don't know what format Firebase Auth uses for the JWT it gives you, but regardless, it won't work with Directus, as the token would be signed by a different party (so doesn't share the same encryption key). 

I'd recommend setting up Firebase as an oAuth provider: https://docs.directus.io/guides/api-config/#oauth-single-sign-on-sso-openid

**Reply by dukesx** [reply]
I checked, Firebase Auth cant work with Grant. Anyway, i just hope someday Directus integrates with flutter.

**Reply by rijkvanzanten** [reply]
Why not? If Firebase follows the oAuth spec, you can use it with grant :)

**Comment by romainbriand** [comment]
@rijkvanzanten ~~Actually what is missing is the ability to implement a custom authentication middleware, IMO.
https://github.com/directus/directus/blob/b84d3be9814e6bb04d7b87d5f99ee07910f09469/api/src/middleware/authenticate.ts is totally locked to the Directus JWT.~~

~~I think we should be able to sign requests with Firebase JWT tokens and the middleware to say YES or NO by checking the JWT token with Firebase token verification service. The logic then stays the same, retrieves the user and voila.~~

~~Is there any plan to add an ability to override authentication middleware? That would be a great help, I must say that it is the only missing thing that holds me to migrate to Directus 😢~~

~~Any suggestion welcome!~~

**Reply by rijkvanzanten** [reply]
> I think we should be able to sign requests with Firebase JWT tokens and the middleware to say YES or NO by checking the JWT token with Firebase token verification service.

Wouldn't that add a lot of overhead? You'd end up hitting the firebase service every request 🤔 

I would imagine you'd use the Firebase JWT to "login" to Directus, similar to how the oAuth flow lets you enter with a token coming from the service provider 🤔

**Reply by romainbriand** [reply]
@rijkvanzanten My bad, I had a bad understanding and messed up between login and authentication! Thanks for your reply, I should get some rest! 😅

**Reply by rijkvanzanten** [reply]
Gotcha! You could already create a custom endpoint to do that then 🙂 You can use the authentication service in a custom endpoint to login as any user based on email, so if you check the JWT with firebase first, you can take it from there 🙂

**Reply by rijkvanzanten** [reply]
Something like (pseudo code)

```ts
export function registerRoutes({ services }) {
  const { AuthenticationService } = services;

  router.get('/', async (req, res) => {
    const service = new AuthenticationService({ schema: req.schema });
    
    const { email } = await getEmailFromFirebase(req.params.firebaseToken);

    const { accessToken, refreshToken, expires } = await AuthenticationService.authenticate({ email });

    res.json({ accessToken, refreshToken, expires });
  });
});
```

**Reply by romainbriand** [reply]
That could do it but unfortunately Firebase Auth accounts doesn't rely on email to make unicity but on a UID. There are many ways a user authenticated with Firebase doesn't have an email associated to its account (ie: he used phone number auth, anonymous auth, Facebook auth without sharing email,...). As email is a mandatory option for the `authenticate` method, it feels like impossible to use in a reliable fashion for this situation.

**Reply by romainbriand** [reply]
Also, I'm afraid Grant/OAuth approach isn't an option either because it feels like relying on email based authentication in the end as well: https://github.com/directus/directus/blob/main/api/src/controllers/auth.ts#L312-L320

**Reply by rijkvanzanten** [reply]
You could add a "firebase_uid" field to directus_users, then lookup the associated account by firebase id, and then login as that user 🤔

**Reply by romainbriand** [reply]
Login relies on authentication service and authentication service uses hardcoded `email` field lookup, right? I don't know all the internals of Directus but that's what it seems to be from the source code: https://github.com/directus/directus/blob/main/api/src/services/authentication.ts#L62-L66 so I don't get how I could do it with another field and get `{ accessToken, refreshToken, expires }` after that.
I'm probably missing something 🤔

**Reply by rijkvanzanten** [reply]
I meant you could read the user record first (based on the firebase id), and then use that users email to go through the regular authenticate step 👍🏻

**Reply by romainbriand** [reply]
Perfect, I see now even though I think at least the field should become customizable or get the ability to put a custom authenticator would be nice feature. Firebase Auth is widely used 🙂
Thanks for the help and keep up the good work guys!

**Reply by rijkvanzanten** [reply]
Totally! The same goes for generic LDAP / Azure Active Directory / etc 🙂 

The first order of business is making the current "simple" setup work well enough for most use cases, and then look into how we can "replace" Directus Users with an external service. The most interesting thing will be trying to figure out how/where to store the "critical" user data vs the meta user data (eg how do you tie a firebase user to a directus role)

**Reply by yfi** [reply]
> Login relies on authentication service and authentication service uses hardcoded `email` field lookup, right? I don't know all the internals of Directus but that's what it seems to be from the source code: https://github.com/directus/directus/blob/main/api/src/services/authentication.ts#L62-L66 so I don't get how I could do it with another field and get `{ accessToken, refreshToken, expires }` after that. I'm probably missing something 🤔

The email field in the user object doesn't seem to enforce an email address and will accept any string. So you could create directus users with the firebase uid in the email field and a generated password/token. Store that password/token in a firebase doc  then use it to login to directus once authenticated into fb.  I'm not sure if there are any security concerns with that, but this is my current plan of action.

**Reply by romainbriand** [reply]
@yfi sounds smart 💡 though I don't really like this kind of hacky ways. It may work for now but you could get locked to an outdated version if some more validation comes on the Directus side of code.

**Reply by benhaynes** [reply]
Part of what we're doing in the new auth flow is allowing _any_ field to be the user identifier. So instead of the `directus_user.email` field, you could add and use a `username`, or anything else. Right, @rijkvanzanten and @aidenfoxx?

**Reply by rijkvanzanten** [reply]
For the local auth flow, yeah 👍🏻

**Reply by romainbriand** [reply]
> Part of what we're doing in the new auth flow is allowing _any_ field to be the user identifier. So instead of the `directus_user.email` field, you could add and use a `username`, or anything else. Right, @rijkvanzanten and @aidenfoxx?

This is very good news!

**Reply by yfi** [reply]
brilliant. keep doing the good work. thanks.

**Reply by aidenfoxx** [reply]
@rijkvanzanten Actually, you asked me to remove the stuff that allowed logging in with an alternate identifier for local. But it's easy enough to add back in. 😅 

The new OAuth2 and OpenID flows do allow using an alternative identifier though. So you would just specify `AUTH_<provider>_IDENTIFIER_KEY="username"` in the config, and it would use the "username" field from the OAuth response to identify the user.

**Reply by rijkvanzanten** [reply]
> @rijkvanzanten Actually, you asked me to remove the stuff that allowed logging in with an alternate identifier for local. But it's easy enough to add back in. 😅

Yeah, it's part of the bigger sprint on auth, but not currently included 🙂 Once we have the dynamic login page, we can officially start documenting configuring more than 1 local driver, at which point the custom identifier becomes important and available

**Reply by aidenfoxx** [reply]
Perfect. Just didn't want anyone thinking it was available right now. 😄

**Comment by carolin-skiply** [comment]
Hi guys,
Is it now possible to authenticate with Firebase? If yes, how? is there a documentaiton?
Thanks,

**Reply by aidenfoxx** [reply]
You should be able to configure it using OpenID: https://firebase.google.com/docs/auth/web/openid-connect

See Directus docs for config options: https://docs.directus.io/self-hosted/sso.html

**Comment by reev-create** [comment]
Yes, I managed to get endpoints work for us to make Firebase as an authentication provider. If you want I could share it with you.

**Reply by carolin-skiply** [reply]
Yes please!

**Reply by maxmblack** [reply]
I would be interested too! 👍

**Reply by ninogjoni** [reply]
Would also be interested!

**Reply by aidenfoxx** [reply]
It should be as simple as:

```
AUTH_PROVIDERS="firebase"

AUTH_FIREBASE_DRIVER="openid"
AUTH_FIREBASE_CLIENT_ID="..."
AUTH_FIREBASE_CLIENT_SECRET="..."
AUTH_FIREBASE_ISSUER_URL="https://securetoken.google.com/<your_project_id>/.well-known/openid-configuration"
AUTH_FIREBASE_IDENTIFIER_KEY="email"
```

**Reply by GustavoBlaze** [reply]
how did you do that?

**Reply by sorin89** [reply]
@reev-create please provide an example if possible. I'm not sure where to get the AUTH_FIREBASE_CLIENT_ID and AUTH_FIREBASE_CLIENT_SECRET in the example above

**Comment by MichaelFBA** [comment]
I receive an error about firebase not supporting code flow `OpenID provider does not support required code flow`
eg. https://securetoken.google.com/next-test-c9c0b/.well-known/openid-configuration

**Reply by aidenfoxx** [reply]
I'm not able to test Firebase myself, but after googling it appears you have to set your OIDC config to "Code Flow" in the identity provider settings.

**Reply by MichaelFBA** [reply]
Yes, there is a custom open id code flow but it asks for an issuer url. I imagine this is for something like Auth0.

**Reply by aidenfoxx** [reply]
That sounds like Firebase would be the authentication client, rather than the server in that case. Unfortunately, I'm not familiar enough with Firebase to provide additional support.

**Reply by MichaelFBA** [reply]
Yes exactly. No problem.

**Reply by ninogjoni** [reply]
Did you accomplish it @MichaelFBA? Could you tell us the steps? I am also trying to include those files from the following link: https://gist.github.com/cotyhamilton/14addb2be92fe7829dcdef0bf5eea10d

**Reply by MichaelFBA** [reply]
No didn't manage sorry

**Reply by MichaelFBA** [reply]
We ended up just using google

**Comment by LeonManolo** [comment]
Did anyone get it to work?
